x-pack/legacy/plugins/infra/server/lib/adapters/metadata/elasticsearch_metadata_adapter.ts

```diff
@@ -73,6 +82,134 @@ export class ElasticsearchMetadataAdapter implements InfraMetadataAdapter {
       buckets,
     };
   }
+
+  public async getCloudMetricMetadata(
+    req: InfraFrameworkRequest,
+    sourceConfiguration: InfraSourceConfiguration,
+    instanceId: string
+  ): Promise<InfraCloudMetricsAdapterResponse> {
+    const metricQuery = {
+      allowNoIndices: true,
+      ignoreUnavailable: true,
+      index: sourceConfiguration.metricAlias,
+      body: {
+        query: {
+          bool: {
+            filter: [{ match: { 'cloud.instance.id': instanceId } }],
+            should: CLOUD_METRICS_MODULES.map(module => ({ match: { 'event.module': module } })),
+          },
+        },
+        size: 0,
+        aggs: {
+          metrics: {
+            terms: {
+              field: 'event.dataset',
+              size: 1000,
+            },
+          },
+        },
+      },
+    };
+
+    const response = await this.framework.callWithRequest<
+      {},
+      {
+        metrics?: InfraMetadataAggregationResponse;
+      }
+    >(req, 'search', metricQuery);
+
+    const buckets =
+      response.aggregations && response.aggregations.metrics
+        ? response.aggregations.metrics.buckets
+        : [];
+
+    return { buckets };
+  }
+
+  public async getNodeInfo(
+    req: InfraFrameworkRequest,
+    sourceConfiguration: InfraSourceConfiguration,
+    nodeId: string,
+    nodeType: InfraNodeType
+  ): Promise<InfraNodeInfo> {
+    // If the nodeType is a Kubernetes pod then we need to get the node info
+    // from a host record instead of a pod. This is due to the fact that any host
+    // can report pod details and we can't rely on the host/cloud information associated
+    // with the kubernetes.pod.uid. We need to first lookup the `kubernetes.node.name`
+    // then use that to lookup the host's node information.
+    if (nodeType === InfraNodeType.pod) {
+      const kubernetesNodeName = await this.getPodNodeName(
+        req,
+        sourceConfiguration,
+        nodeId,
+        nodeType
+      );
+      if (kubernetesNodeName) {
+        return this.getNodeInfo(req, sourceConfiguration, kubernetesNodeName, InfraNodeType.host);
+      }
+      return {};
+    }
+    const params = {
+      allowNoIndices: true,
+      ignoreUnavailable: true,
+      terminateAfter: 1,
+      index: sourceConfiguration.metricAlias,
+      body: {
+        size: 1,
+        _source: ['host.*', 'cloud.*'],
```

Comment:
Can you explain what this does? (Not a critique, I'm still learning ES query details.)

Reply:
That controls which fields are returned in the source. For this example only the top level `host` and `cloud` objects are returned in the source document.

Reply:
Why?

Reply:
It's an optimization, we only need the `host` and `cloud` objects, this filters out all the other fields an only transfers the ones we care about. When the object is returned it's in the exact format and we need.

Reply:
Thanks for the explanation. Would other queries in the Infra UI also benefit from this type of optimization? (This is totally unrelated to this change of course.)